feat(github-actions): add compare link to PR body definitions

kopach opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

Added a `prBodyDefinitions` field to the GitHub Actions manager under `defaultConfig`. This field includes a `Change` template that generates a direct compare link for digest updates in PR descriptions.

- https://github.com/renovatebot/renovate/discussions/35605

## Context

This change improves the PR descriptions for GitHub Actions updates by providing a direct link to compare changes between the old and new digests. This enhancement aligns with the existing behavior for npm updates and makes it easier for reviewers to understand the changes.

Closes #issue_number (if applicable).

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

kopach committed:
feat(github-actions): add compare link to PR body definitions

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/renovatebot/renovate?pullRequest=35613) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/renovatebot/renovate?pullRequest=35613) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/renovatebot/renovate?pullRequest=35613) it.</sub>

rarkins reviewed (changes requested):
I'd like tests so we can make sure this doesn't break any scenarios

kopach commented:
hi @rarkins, I have very little knowledge about this project. Could you help me to understand where and how to write tests for this feature. If you could point me to a test case for a similar feature - that would be helpful. Thanks in advance

kopach commented:
hi @viceice, @rarkins. I've updated a bit the rule. Added also check for `sourceUrl`. If it's there - I'm using it (in my cases it's github anyway).
Also here is a public repo, where this rule works https://github.com/kopach/renovate-pull-35613
The setup is following:
https://github.com/kopach/renovate-pull-35613/blob/1b2396778c1da7d5a43372ede42460c3f25f3087/.github/renovate.json5#L4-L11
Here is result PR https://github.com/kopach/renovate-pull-35613/pull/27
**Note** `matchDepTypes: ['action']`. I've tried to work around how to make this also for other types - but couldn't find reliable source of diff changes e.g. https://github.com/kopach/renovate-pull-35613/pull/26 would require checking diff for node itself which is unavailable in template

P.S. I believe there is a bug in `engines` type. [See this PR](https://github.com/kopach/renovate-pull-35613/pull/25), where comparison leads to https://renovatebot.com/diffs/npm/node/v22.13.1/v22.17.0 which never resolves

kopach commented on lib/modules/manager/github-actions/index.ts:
```diff
@@ -13,6 +13,10 @@ export const defaultConfig = {
   ],
+  prBodyDefinitions: {
+    Change:
+      '[`{{#if displayFrom}}{{{displayFrom}}}{{else}}{{{currentValue}}}{{/if}}` -> `{{#if displayTo}}{{{displayTo}}}{{else}}{{{newValue}}}{{/if}}`]({{#if depName}}https://github.com/{{replace "/" "%2F" depName}}/compare/{{#if currentDigest}}{{{currentDigestShort}}}{{else}}{{{currentValue}}}{{/if}}...{{#if newDigest}}{{{newDigestShort}}}{{else}}{{{newValue}}}{{/if}}{{/if}})',
```
Isn't this file specifically for GitHub Actions, as the name suggests? I'm not very familiar with this repo, so please feel free to update this PR as needed. For me, keeping this in repo's renovate config is OK, but I'm happy to push it upstream and make it available for everyone

viceice commented on lib/modules/manager/github-actions/index.ts:
```diff
@@ -13,6 +13,10 @@ export const defaultConfig = {
   ],
+  prBodyDefinitions: {
+    Change:
+      '[{{#if displayFrom}}{{{displayFrom}}}{{else}}{{{currentValue}}}{{/if}} -> {{#if displayTo}}{{{displayTo}}}{{else}}{{{newValue}}}{{/if}}]({{#if sourceUrl}}{{{sourceUrl}}}/compare/{{#if currentDigest}}{{{currentDigestShort}}}{{else}}{{{currentValue}}}{{/if}}...{{#if newDigest}}{{{newDigestShort}}}{{else}}{{{newValue}}}{{/if}}{{else}}{{#if depName}}https://github.com/{{replace "/" "%2F" depNa ...
```
```suggestion
      '[{{#if displayFrom}}{{{displayFrom}}}{{else}}{{{currentValue}}}{{/if}} -> {{#if displayTo}}{{{displayTo}}}{{else}}{{{newValue}}}{{/if}}]({{#if sourceUrl}}{{{sourceUrl}}}/compare/{{#if currentDigest}}{{{currentDigestShort}}}{{else}}{{{currentValue}}}{{/if}}...{{#if newDigest}}{{{newDigestShort}}}{{else}}{{{newValue}}}{{/if}}{{else}}{{#if depName}}https://github.com/{{depName}}/compare/{{#if currentDigest}}{{{currentDigestShort}}}{{else}}{{{currentValue}}}{{/if}}...{{#if newDigest}}{{{newDigestShort}}}{{else}}{{{newValue}}}{{/if}}{{/if}}{{/if}})',
```

The `%2F` encoding is wrong for github. Remove the github fallback because all actions will have a source url.
If they don't have, then github is wrong too.

viceice reviewed (commented):
use the `digest` child object like bun. it makes it more simple

https://github.com/renovatebot/renovate/blob/db9eef8dc62e30a48afb769f74c55ea370d39455/lib/modules/manager/bun/index.ts#L17-L22

kopach committed:
Update lib/modules/manager/github-actions/index.ts

Co-authored-by: Michael Kriese <michael.kriese@gmx.de>

kopach commented:
> use the `digest` child object like bun. it makes it more simple
> 
> https://github.com/renovatebot/renovate/blob/db9eef8dc62e30a48afb769f74c55ea370d39455/lib/modules/manager/bun/index.ts#L17-L22

You could reply with something like this:

Hi @viceice, thanks for your review. If my understanding is correct, this `digest` property is only for Renovate's "digest" changes. In this case, we also expect `major`/`minor`/`patch` changes. My quick testing shows that `newDigestShort`/ `currentDigestShort` is also available for non-digest changes, and that's what we most probably want here, since the version is not necessary will be exact in the package release log (there could be even no tags in package repo with the exact version, but the hash will be there).

That’s why I would keep the logic outside of the digest. 

P.S.
working with those 1 line handlebars syntax is quite complex, have you thought on multiline support like this
```hbs
[
  {{#if displayFrom}}
    {{{displayFrom}}}
  {{else}}
    {{{currentValue}}}
  {{/if}}
  ->
  {{#if displayTo}}
    {{{displayTo}}}
  {{else}}
    {{{newValue}}}
  {{/if}}
](
  {{#if sourceUrl}}
    {{{sourceUrl}}}/compare/
    {{#if currentDigest}}
      {{{currentDigestShort}}}
    {{else}}
      {{{currentValue}}}
    {{/if}}
    ...
    {{#if newDigest}}
      {{{newDigestShort}}}
    {{else}}
      {{{newValue}}}
    {{/if}}
  {{else}}
    {{#if depName}}
      https://github.com/{{depName}}/compare/
      {{#if currentDigest}}
        {{{currentDigestShort}}}
      {{else}}
        {{{currentValue}}}
      {{/if}}
      ...
      {{#if newDigest}}
        {{{newDigestShort}}}
      {{else}}
        {{{newValue}}}
      {{/if}}
    {{/if}}
  {{/if}}
)
```

RahulGautamSingh reviewed (approved):
Bit long, but seems fine,

jamietanna commented:
Hey @kopach apologies for the late reply, I think we might have already done this in https://github.com/renovatebot/renovate/pull/39650, where it'll add the `(changelog)` link

If you'd like to refactor this PR to add the `Change` column under that preset, that'd probably simplify what we're doing here 🤞
